### ISSUE TITLE
Add startSqueezing/endSqueezing/simulateSqueeze

### ISSUE THREAD

**Issue by Artyom17** [immersive-web/webxr-test-api#49]
To match the 'select' event testing, why don't we add the startSqueezing/endSqueezing/simulateSqueeze methods, as well as squeezingStarted, squeezingClicked states?

**Comment by Manishearth** [comment]
Yeah, we should add these

**Comment by alcooper91** [comment]
I'm a little hesitant on this. The select button is required on an XRInputSource, and as such we didn't define the select button in [FakeXrButtonType](https://immersive-web.github.io/webxr-test-api/#enumdef-fakexrbuttontype). However, the grip does exist in [FakeXrButtonType](https://immersive-web.github.io/webxr-test-api/#enumdef-fakexrbuttontype), so these would really just be a convenience wrapper around [updateButtonState](https://immersive-web.github.io/webxr-test-api/#dom-fakexrinputcontroller-updatebuttonstate)

**Comment by Manishearth** [comment]
It's possible for an XR device to support squeeze but not a gamepad (we're doing this for hands on Hololens)

**Comment by alcooper91** [comment]
updateButtonState doesn't require gamepad though?

It still feels wrong to expose two ways of doing the same thing

**Comment by Manishearth** [comment]
Oh, I see. Yeah.

However: I don't think the updateButtonState can simulate cancellation; which is possible for both select and squeeze events (and leads to a selectend event not paired with a select event). Not a big deal, but something to think about.

**Comment by alcooper91** [comment]
I don't think our mechanism for select can simulate cancellation right now?

we have startSelection/endSelection and then simulateSelect, which IIRC from the explainer is used to simulate either a subframe select or just a convinience method for calling

startSelection
endSelection

**Comment by alcooper91** [comment]
One compromise here could be to allow the webxr_util class to polyfill in such a helper. (at least for the start/stop)

**Comment by Manishearth** [comment]
Hmm, we perhaps should change the APIs then to startSelect/endSelect/cancelSelect?

That maps more cleanly to the "start, end, cancel" framing that's in the spec: https://immersive-web.github.io/webxr/#primary-squeeze-action 

IMO this API should hook directly into such spec algorithms (which are triggered by the user) as much as possible

**Comment by Manishearth** [comment]
In the current test API this hooks more directly into the user side rather than the device side.

**Comment by Manishearth** [comment]
Also, if we're doing this, it might be worth having just startSelect/endSelect/cancelSelect with a `squeeze` optional boolean argument

**Comment by alcooper91** [comment]
That kind of feels like overriding the existing method for the sake of not adding a new one, in a way that feels neither scalable nor intuitive IMO

**Comment by Manishearth** [comment]
That's fair. We can also add a new one then.

**Comment by alcooper91** [comment]
If we're okay with removing one (that we maybe keep polyfilled in) We could also do something like
```
startPrimaryAction(FakeXrPrimaryAction action)
stopPrimaryAction(FakeXrPrimaryAction action)
```

where currently
```
enum FakeXrPrimaryAction {
  "select",
  "squeeze",
}
```

**Comment by alcooper91** [comment]
Though "squeeze" would still be possible via updateButtonState, unless we wanted to explicitly disallow that.

**Comment by Artyom17** [comment]
So, should the implementation of updateButtonState check, that the button[1] has been changed, therefore fire the squeeze event? Currently, it just updates the gamepad object (in Chromium implementation) and this doesn't and wouldn't cause the squeeze events firing.

Also, should the updating button[0] cause select events to fire somehow?

**Comment by alcooper91** [comment]
The goal of the Test API has been to mimic a "real" device. updateButtonState is our way of saying "Hey, when you next check for data pretend that the state of this button has changed"
Currently in Chrome, we only update the Gamepad object, because we didn't have any other mechanism of passing that data into blink.

Updating button[0] (since this is the gamepad button that maps to the select event) *should* fire that event. However, the way that we have currently designed the API (taking a [FakeXrButtonType](https://immersive-web.github.io/webxr-test-api/#enumdef-fakexrbuttontype) which doesn't include "select"), makes it impossible to update the select state via "updateButtonState"

**Comment by Manishearth** [comment]
The enum solution sounds good to me!

**Comment by alcooper91** [comment]
Actually, that still leaves us with two ways to press some buttons.
What if we:
* remove FakeXRInputSourceInit selectionStarted/selectionClicked
* move startSelection/endSelection/simulateSelection to be poly-filled in
* Add "select" to FakeXRButtonType

**Comment by Manishearth** [comment]
The thing that gives me pause is that the FakeXRButton stuff contains a lot of gamepad stuff as well, and ideally that would be separate (we don't implement it).


But otherwise this seems fine. I'm also okay with the compat breaking changes, really.
> to make it more generic for Hands and other forms of input.

The hands api isn't fixed enough for this to make sense, yet.

**Comment by alcooper91** [comment]
> The thing that gives me pause is that the FakeXRButton stuff contains a lot of gamepad stuff as well, and ideally that would be separate (we don't implement it).

I don't think that would be a major issue. Your implementation can simply ignore buttons it doesn't know/doesn't want to handle. Really the only place that the state of those other buttons *should* bubble up would be via the Gamepad/gamepad tests. So as long as you just ignore the values, the other tests should be fine?

> > to make it more generic for Hands and other forms of input.
> 
> The hands api isn't fixed enough for this to make sense, yet.

While I understand that, if we were to bubble up any additional events, they would likely be considered an "Input Action" hence my proposal to name them FakeXRActionType rather than FakeXRButtonType

I don't believe we can do partial enums, but I think it would be fine to add spec text to say that certain enum values may not be supported if they are required by a module that the UA doesn't support.
Alternatively, we could do a similar approach to the "feature-descriptors" that the spec uses (though maybe enforce strings), and give "valid" strings on a per-module basis.

**Comment by Manishearth** [comment]
> I don't believe we can do partial enums, but I think it would be fine to add spec text to say that certain enum values may not be supported if they are required by a module that the UA doesn't support.

This is fine.

> While I understand that, if we were to bubble up any additional events, they would likely be considered an "Input Action" hence my proposal to name them FakeXRActionType rather than FakeXRButtonType

I believe the current thinking is that hands will also have select and squeeze events, and nothing more, everything else is exposed as state and not events.
